CONTRIBUTING.rst

```diff
@@ -0,0 +1,500 @@
+rebasing easier, and keeps the change history clear and clean.
+
+When contributing to the Zephyr Project, it is also important you provide as much
+information as you can about your change, update appropriate documentation,
+and test your changes thoroughly before submitting.
+
+The general GitHub workflow used by Zephyr developers uses a combination of
+command line git commands and browser interaction with GitHub.  As it is with
+git, there are multiple ways of getting a task done.  We'll describe a typical
+workflow here:
+
+.. _Create a Fork of Zephyr:
+   https://github.com/zephyrproject-rtos/zephyr#fork-destination-box
+
+#. `Create a Fork of Zephyr`_
+   to your personal account on GitHub
+
+#. On your development computer, clone the fork you just made::
+
+     $ git clone https://github.com/<your github id>/zephyr
+
+   This would be a good time to let git know about the upstream repo too::
+
+     $ git remote add upstream https://github.com/zephyrproject-rtos/zephyr.git
+
+   and verify the remote repos::
+
+     $ git remote -v
+
+#. Create a topic branch (off of master) for your work (if you're addressing
+   Jira issue, we suggest including the Jira issue number in the branch name)::
+
+     $ git checkout master
+     $ git checkout -b fix_comment_typo
+
+   Some Zephyr subsystems do development work on a separate branch from
+   master so you may need to indidicate this in your checkout::
+
+     $ git checkout -b fix_out_of_date_patch origin/net
+
+#. Make changes, test locally, change, test, test again, ...  (Check out the
+   prior chapter on `sanitycheck`_ as well).
+
+#. When things look good, start the pull-request process by adding your changed
+   files::
+
+     $ git add [file(s) that changed, add -p if you want to be more specific]
+
+   You can see files that are not yet staged using::
+
+     $ git status
+
+#. Verify changes to be committed look as you expected::
+
+     $ git diff --cached
+
+#. Commit your changes to your local repo::
+
+     $ git commit -s
+
+   The ``-s`` option automatically adds your ``Signed-off-by:`` to your commit
+   message.  Your commit will be rejected without this line that indicates your
+   agreement with the `DCO`_.  See the `Commit Guidelines`_ section for
+   specific guidelines for writing your commit messages.
+
+
+#. Push your topic branch with your changes to your fork in your personal
+   GitHub account::
+
+     $ git push origin fix_comment_typo
+
+   or if working on a different branch than ``master`` (as mentioned above)::
+
+     $ git push origin/net fix_out_of_date_patch
+
+#. In your web browser, go to your forked repo and click on the
+   ``Compare & pull request`` button for the branch you just worked on and
+   you want to open a pull request with.
+
+#. Review the pull request changes, and verify that you are opening a
+   pull request for the appropriate branch. The title and message from your
+   commit message should appear as well.
```

Comment:
This workflow assumes that the pull request contains exactly one commit, which isn't the common practice and goes against the principle of making a series of small changes that are easier to rebase etc.